nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2464 @@
+    let osagactx = sagactx.user_data();
+    let RealParams { serialized_authn, authz_instance, ref update, .. } =
+        sagactx.saga_params::<RealParams>()?;
+    let lock = sagactx.lookup::<instance::UpdaterLock>(INSTANCE_LOCK)?;
+
+    let opctx =
+        crate::context::op_context_for_saga_action(&sagactx, &serialized_authn);
+    let log = osagactx.log();
+    let nexus = osagactx.nexus();
+
+    let instance_id = authz_instance.id();
+
+    debug!(
+        log,
+        "instance update: committing new runtime state and unlocking...";
+        "instance_id" => %instance_id,
+        "new_runtime" => ?update.new_runtime,
+        "lock" => ?lock,
+    );
+
+    let did_unlock = osagactx
+        .datastore()
+        .instance_commit_update(
+            &opctx,
+            &authz_instance,
+            &lock,
+            &update.new_runtime,
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    info!(
+        log,
+        "instance update: committed update new runtime state!";
+        "instance_id" => %instance_id,
+        "new_runtime" => ?update.new_runtime,
+        "did_unlock" => ?did_unlock,
+    );
+
+    if update.network_config.is_some() {
+        // If the update we performed changed networking configuration, activate
+        // the V2P manager and VPC router RPWs, to ensure that the V2P mapping
+        // and VPC for this instance are up to date.
+        //
+        // We do this here, rather than in the network config update action, so
+        // that the instance's state in the database reflects the new rather
+        // than the old state. Otherwise, if the networking RPW ran *before*
+        // writing the new state to CRDB, it will run with the old VMM, rather
+        // than the new one, and probably do nothing. Then, the networking
+        // config update would be delayed until the *next* background task
+        // activation. This way, we ensure that the RPW runs *after* we are in
+        // the new state.
+
+        nexus.background_tasks.task_v2p_manager.activate();
+        nexus.vpc_needed_notify_sleds();
+    }
+
+    // Check if the VMM or migration state has changed while the update saga was
+    // running and whether an additional update saga is now required. If one is
+    // required, try to start it.
+    //
+    // TODO(eliza): it would be nice if we didn't release the lock, determine
+    // the needed updates, and then start a new start-instance-update saga that
+    // re-locks the instance --- instead, perhaps we could keep the lock, and
+    // try to start a new "actual" instance update saga that inherits our lock.
+    // This way, we could also avoid computing updates required twice.
+    // But, I'm a bit sketched out by the implications of not committing update
+    // and dropping the lock in the same operation. This deserves more thought...
+    if let Err(error) =
+        chain_update_saga(&sagactx, authz_instance, serialized_authn).await
```

Comment:
Since "chaining" hands off the lock, it's critically important that we don't perform any operations *after* the chained sub-saga has started, right?

E.g., I imagine it would be "real bad" if we thought we handed off the lock, but then unwound this update saga because of an error while that chained subsaga is running

Reply:
It's not actually that bad: we have released the lock here, and the new child saga will generate its own ID and acquire the lock again. Since unlocking is idempotent, if we were to add a fallible operation after this and unwind, the unwinding saga would attempt to unlock the instance either (a) while the instance is unlocked, which doesn't do anything, or (b) while the instance is locked by the *new* saga, which...also doesn't do anything. So it's fine, actually.

Reply:
ahhhh, got it.

My confusion here stemmed from an assumptions that:

1. holding the instance lock gives us the ability to mutate the instance record
2. we may mutate the instance record on both "actions" and "undo" actions
3. The "chain" operation performs lock handoff

If all of these are true, I was worried that we'd hand-off the lock in (3), but then possibly unwind and mutate the instance record because of (2). After all, this would be "undoing" all the actions we took during the saga **before** we get back to the `siu_unbecome_updater` node.

However, re-reading the instance_update actions, I see there aren't really "undo actions" that mutate the instance.

If there **were** undo actions that could mutate instance state, this would be a problem, right?

(Agreed that it isn't an issue right now though)

Reply:
> However, re-reading the instance_update actions, I see there aren't really "undo actions" that mutate the instance.
> 
> If there **were** undo actions that could mutate instance state, this would be a problem, right?

Yes, that's correct. This saga is a bit different because, unlike `instance-start`, `instance-migrate`, or `instance-delete`, the `instance-update` saga is not attempting to *perform a state change*, it is attempting to *update the database (and stuff) to match a state change that has already occurred*. 

If, for example, we run an `instance-start` saga, and the instance cannot actually be started, of course we want to undo any database changes we've made, because the instance was not actually started. On the other hand, if an instance's VMM shuts down and we start an `instance-update` saga to move it to the `Destroyed` state...even if some action along the way fails, the instance _is_ still `Destroyed`; that state transition has *already happened* and unwinding the saga cannot and should not un-destroy it. So, unlike other sagas, we want to leave basically anything we've successfully done in place when unwinding. If there was something we weren't able to do, one of the instance-update-related RPWs will start a new update saga to try it again.

Reply:
Now that I think about it, perhaps this ought to go in my Big Doc Comment...

Reply:
I added an explanation of the unwinding behavior in https://github.com/oxidecomputer/omicron/pull/5749/commits/afd9cf147c53c8bd2489f50ad8ef93e8387cddec, what do you think?